pkg/querier/queryrange/roundtrip.go

```diff
@@ -55,60 +54,87 @@ func NewTripperware(cfg Config, log log.Logger, limits Limits, registerer promet
-			if logSelector, ok := expr.(logql.LogSelectorExpr); ok {
-				if err := validateLimits(req, params, limits); err != nil {
-					return nil, err
-				}
-
-				// backport the old regexp params into the query params
-				regexp := params.Get("regexp")
-				if regexp != "" {
-					logSelector = logql.NewFilterExpr(logSelector, labels.MatchRegexp, regexp)
-					params.Set("query", logSelector.String())
-					req.URL.RawQuery = params.Encode()
-				}
-				filter, err := logSelector.Filter()
-				if err != nil {
-					return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
-				}
-				if filter != nil {
-					return logFilterRT.RoundTrip(req)
-				}
-			}
-			return next.RoundTrip(req)
-		})
+		return newRoundTripper(next, logFilterRT, metricRT, limits)
 	}, cache, nil
 }
 
-// validates log entries limits
-func validateLimits(req *http.Request, params url.Values, limits Limits) error {
-	userID, err := user.ExtractOrgID(req.Context())
+type roundTripper struct {
+	next, log, metric http.RoundTripper
+
+	limits Limits
+}
+
+// newRoundTripper creates a new queryrange roundtripper
+func newRoundTripper(next, log, metric http.RoundTripper, limits Limits) roundTripper {
+	return roundTripper{
+		log:    log,
+		limits: limits,
+		metric: metric,
+		next:   next,
+	}
+}
+
+func (r roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if !strings.HasSuffix(req.URL.Path, "/query_range") && !strings.HasSuffix(req.URL.Path, "/prom/query") {
+		return r.next.RoundTrip(req)
+	}
+	err := req.ParseForm()
 	if err != nil {
-		return httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
 	}
+	rangeQuery, err := loghttp.ParseRangeQuery(req)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+	expr, err := logql.ParseExpr(rangeQuery.Query)
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+	switch e := expr.(type) {
+	case logql.SampleExpr:
+		return r.metric.RoundTrip(req)
+	case logql.LogSelectorExpr:
+		filter, err := transformRegexQuery(req, e).Filter()
+		if err != nil {
+			return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+		}
+		if err := validateLimits(req, rangeQuery.Limit, r.limits); err != nil {
+			return nil, err
+		}
+		if filter == nil {
+			return r.next.RoundTrip(req)
+		}
+		return r.log.RoundTrip(req)
 
-	reqLimit, err := strconv.Atoi(params.Get("limit"))
+	default:
+		return r.next.RoundTrip(req)
+	}
+}
+
+// transformRegexQuery backport the old regexp params into the v1 query format
+func transformRegexQuery(req *http.Request, expr logql.LogSelectorExpr) logql.LogSelectorExpr {
+	regexp := req.Form.Get("regexp")
+	if regexp != "" {
+		expr = logql.NewFilterExpr(expr, labels.MatchRegexp, regexp)
+		params := req.URL.Query()
+		params.Set("query", expr.String())
+		req.URL.RawQuery = params.Encode()
+		// force the form and query to be parsed again.
+		req.Form = nil
```

Comment:
I'm a bit scared of setting this to nil here. I'm afraid it exposes us to issues extending this in the future -- it's easy to think the form would be parsed already. That being said, I think we should merge this. It's just something we should think about improving.

Reply:
Yeah I don't have a solution yet for that. The problem is that ParseForm does nothing if it's already populated and we need to update the querystring.

Reply:
If you remove this a test will fail.